File: blik-base/src/main/java/com/adyen/checkout/blik/BlikOutputData.java
/*
 * Copyright (c) 2020 Adyen N.V.
 *
 * This file is open source and available under the MIT license. See the LICENSE file for more info.
 *
 * Created by josephj on 4/12/2020.
 */

package com.adyen.checkout.blik;

import android.support.annotation.NonNull;
import android.text.TextUtils;

import com.adyen.checkout.base.component.OutputData;
import com.adyen.checkout.base.validation.ValidatedField;

class BlikOutputData implements OutputData {

    private final ValidatedField<String> mBlikCodeField;

    BlikOutputData(@NonNull String blikCode) {
        mBlikCodeField = new ValidatedField<>(blikCode,
                TextUtils.isEmpty(blikCode) ? ValidatedField.Validation.PARTIAL : ValidatedField.Validation.VALID);
    }

    @Override
    public boolean isValid() {
        return mBlikCodeField.isValid();
    }

    @NonNull
    public ValidatedField<String> getBlikCodeField() {
        return mBlikCodeField;
    }
}

File: blik/src/main/java/com/adyen/checkout/blik/internal/ui/model/BlikOutputData.kt
/*
 * Copyright (c) 2020 Adyen N.V.
 *
 * This file is open source and available under the MIT license. See the LICENSE file for more info.
 *
 * Created by josephj on 4/12/2020.
 */
package com.adyen.checkout.blik.internal.ui.model

import com.adyen.checkout.blik.R
import com.adyen.checkout.components.core.internal.ui.model.FieldState
import com.adyen.checkout.components.core.internal.ui.model.OutputData
import com.adyen.checkout.components.core.internal.ui.model.Validation
import com.adyen.checkout.core.old.AdyenLogLevel
import com.adyen.checkout.core.old.internal.util.adyenLog

internal class BlikOutputData(blikCode: String) : OutputData {

    val blikCodeField: FieldState<String>
    override val isValid: Boolean
        get() = blikCodeField.validation.isValid()

    private fun getBlikCodeValidation(blikCode: String): Validation {
        try {
            if (blikCode.isNotEmpty()) blikCode.toInt()
        } catch (e: NumberFormatException) {
            adyenLog(AdyenLogLevel.ERROR, e) { "Failed to parse blik code to Integer" }
            return Validation.Invalid(R.string.checkout_blik_code_not_valid)
        }
        return if (blikCode.length == BLIK_CODE_LENGTH) {
            Validation.Valid
        } else {
            Validation.Invalid(R.string.checkout_blik_code_not_valid)
        }
    }

    companion object {
        private const val BLIK_CODE_LENGTH = 6
    }

    init {
        blikCodeField = FieldState(blikCode, getBlikCodeValidation(blikCode))
    }
}
